internal/daemon/dpudaemon.go

```diff
@@ -125,96 +150,99 @@ func (d *DpuDaemon) Listen() (net.Listener, error) {
 		return d.cniCmdNfDelHandler(r)
 	}
 
-	d.cniserver = cniserver.NewCNIServer(add, del, cniserver.WithSocketPath(d.cniServerPath))
+	d.cniserver = cniserver.NewCNIServer(add, del, cniserver.WithSocketPath(d.pathManager.CNIServerPath()))
 
 	return lis, err
 }
 
 func (d *DpuDaemon) ListenAndServe() error {
-	var wg sync.WaitGroup
-	done := make(chan error, 1)
-
 	listener, err := d.Listen()
 
 	if err != nil {
 		d.log.Error(err, "Failed to listen")
 		return err
 	}
 
-	wg.Add(1)
+	return d.Serve(listener)
+}
+
+func (d *DpuDaemon) Serve(listener net.Listener) error {
+	d.wg.Add(3)
+
 	go func() {
-		d.log.Info("Starging OPI server")
-		if err := d.Serve(listener); err != nil {
-			done <- err
+		d.log.Info("Starting OPI server")
+		d.server = grpc.NewServer()
+		pb.RegisterBridgePortServiceServer(d.server, d)
+		if err := d.server.Serve(listener); err != nil {
+			d.done <- fmt.Errorf("Failed to start serving: %v", err)
 		} else {
-			done <- nil
+			d.done <- nil
 		}
-		wg.Done()
+		d.log.Info("Stopping OPI server")
+		d.wg.Done()
 	}()
 
-	wg.Add(1)
 	go func() {
 		d.log.Info("Starting CNI server")
 		if err := d.cniserver.ListenAndServe(); err != nil {
-			done <- err
+			d.done <- err
 		} else {
-			done <- nil
+			d.done <- nil
 		}
-		wg.Done()
+		d.log.Info("Stopping CNI server")
+		d.wg.Done()
 	}()
 
 	d.setupReconcilers()
-	wg.Add(1)
-
 	ctx, cancelManager := context.WithCancel(ctrl.SetupSignalHandler())
 	go func() {
 		d.log.Info("Starting manager")
-
 		if err := d.manager.Start(ctx); err != nil {
-			done <- err
+			d.done <- err
 		} else {
-			done <- nil
+			d.done <- nil
 		}
-		wg.Done()
+		d.log.Info("Stopping manager")
+		d.wg.Done()
 	}()
 
-	err = <-done
+	d.cancelManager = cancelManager
 
-	cancelManager()
+	err := <-d.done
+	d.cancelManager()
 	d.cniserver.Shutdown(context.TODO())
 	d.server.Stop()
-	wg.Wait()
-
+	d.wg.Wait()
+	d.stopWg.Done()
 	return err
 }
 
-func (d *DpuDaemon) Serve(listen net.Listener) error {
-	d.server = grpc.NewServer()
-	pb.RegisterBridgePortServiceServer(d.server, d)
-	if err := d.server.Serve(listen); err != nil {
-		d.log.Error(err, "Failed to start serving")
-		return err
-	}
-	return nil
+func (d *DpuDaemon) Stop() {
+	d.done <- nil
```

Comment:
close(d.done)

Reply:
This would cause issues when the three go routines write to the channel as they stop. We still want to allow writing to the channel after `stop()` has been called, because that's part of the cleanup.